buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -545,11 +562,65 @@ static int writeUtf8(AbstractByteBuf buffer, int writerIndex, CharSequence seq,
         return writerIndex - oldWriterIndex;
     }
 
+    /**
+     * Returns max bytes length of UTF8 character sequence of the given length.
+     */
+    public static int utf8MaxBytes(final int seqLength) {
+        return seqLength * MAX_BYTES_PER_CHAR_UTF8;
+    }
+
     /**
      * Returns max bytes length of UTF8 character sequence.
+     * <p>
+     * It behaves like {@link #utf8MaxBytes(int)} applied to {@code seq} {@link CharSequence#length()}.
      */
     public static int utf8MaxBytes(CharSequence seq) {
-        return seq.length() * MAX_BYTES_PER_CHAR_UTF8;
+        return utf8MaxBytes(seq.length());
+    }
+
+    /**
+     * Returns the exact bytes length of a well-formed UTF8 character sequence.
+     * <p>
+     * This method is producing a precise estimation only with well-formed UTF8 character sequences, otherwise
+     * it will not perform underrated estimations according to {@link #writeUtf8(ByteBuf, CharSequence)}.
+     */
+    public static int utf8Bytes(final CharSequence seq) {
+        int seqLength = seq.length();
+        //the algorithm is just adding the additional length of not ASCII chars to the baseline ASCII length
+        int encodedLength = seqLength;
+        int i = 0;
+        //ASCII
+        while (i < seqLength && seq.charAt(i) < 0x80) {
+            i++;
+        }
+        if (i < seqLength) {
+            //!ASCII
+            encodedLength += utf8BytesSlowPath(seq, i, seqLength);
+            //integer overflow on utf8BytesSlowPath
+            if (encodedLength < seqLength) {
+                throw new IllegalArgumentException("UTF-8 encoded length overflow");
```

Comment:
Is this something that should ever happen.?

Reply:
`CharSequence` to me means `StringBuilder/CharBuffer/String` as common cases and if you have such big instances you have for sure bigger troubles than just calculating the UTF 8 encoded length :P, but the API doesn't prohibit anyone to provide a custom implementation backed by an arbitrary long stream of chars that could cause that overflow..
If you think that it could be dropped to me it could be safe enough, but if it doesn't cost I will leave it there, wdyt?